zeromq 5 runs on bitbucket pipeline, but 6 doesn't

We have a simple pipeline that installs zeromq with 
- npm i

If we use 5.3.1 version it works fine. If we install the latest:

`npm ERR! code 1`
`npm ERR! path /opt/atlassian/pipelines/agent/build/server/node_modules/zeromq`
`npm ERR! command failed`
`npm ERR! command sh -c -- (shx test -f ./script/build.js || run-s build.js) && cross-env npm_config_build_from_source=true node-gyp-build`
`npm ERR! make: Entering directory '/opt/atlassian/pipelines/agent/build/server/node_modules/zeromq/build'`
`npm ERR!   TOUCH 7722eaf02fcc073df28ffd1fa831e71544edfd42.intermediate`
`npm ERR!   ACTION binding_gyp_libzmq_target_build_libzmq 7722eaf02fcc073df28ffd1fa831e71544edfd42.intermediate`
`npm ERR! Downloading libzmq source from https://github.com/zeromq/libzmq/archive/20de92ac0a2b2b9a1869782a429df68f93c3625e.tar.gz`
`npm ERR! Building libzmq Release`
`npm ERR! cmake -S "libzmq-20de92ac0a2b2b9a1869782a429df68f93c3625e" -B ./build  -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX="/opt/atlassian/pipelines/agent/build/server/node_modules/zeromq/build/libzmq" -DCMAKE_INSTALL_LIBDIR=lib -DBUILD_STATIC=ON -DBUILD_TESTS=OFF -DBUILD_SHARED=OFF -DWITH_DOCS=OFF -DWITH_LIBSODIUM=ON -DWITH_LIBSODIUM_STATIC=ON`
`npm ERR! cmake --build ./build --config Release --target install`
`npm ERR!   TOUCH Release/obj.target/libzmq.stamp`
`npm ERR!   CXX(target) Release/obj.target/zeromq/src/context.o`
`npm ERR! rm 7722eaf02fcc073df28ffd1fa831e71544edfd42.intermediate`
`npm ERR! make: Leaving directory '/opt/atlassian/pipelines/agent/build/server/node_modules/zeromq/build'`
`npm ERR! /lib/x86_64-linux-gnu/libc.so.6: version 'GLIBC_2.33' not found (required by /opt/atlassian/pipelines/agent/build/server/node_modules/zeromq/prebuilds/linux-x64/node.napi.glibc.node)`
`npm ERR! gyp info it worked if it ends with ok`
`npm ERR! gyp info using node-gyp@9.1.0`
`npm ERR! gyp info using node@16.19.1 | linux | x64`
`npm ERR! gyp info find Python using Python version 3.7.3 found at "/usr/bin/python3"`
`npm ERR! gyp http GET https://nodejs.org/download/release/v16.19.1/node-v16.19.1-headers.tar.gz`
`npm ERR! gyp http 200 https://nodejs.org/download/release/v16.19.1/node-v16.19.1-headers.tar.gz`
`npm ERR! gyp http GET https://nodejs.org/download/release/v16.19.1/SHASUMS256.txt`
`npm ERR! gyp http 200 https://nodejs.org/download/release/v16.19.1/SHASUMS256.txt`
`npm ERR! gyp info spawn /usr/bin/python3`
`npm ERR! gyp info spawn args [`
`npm ERR! gyp info spawn args   '/usr/local/lib/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py',`
`npm ERR! gyp info spawn args   'binding.gyp',`
`npm ERR! gyp info spawn args   '-f',`
`npm ERR! gyp info spawn args   'make',`
`npm ERR! gyp info spawn args   '-I',`
`npm ERR! gyp info spawn args   '/opt/atlassian/pipelines/agent/build/server/node_modules/zeromq/build/config.gypi',`
`npm ERR! gyp info spawn args   '-I',`
`npm ERR! gyp info spawn args   '/usr/local/lib/node_modules/npm/node_modules/node-gyp/addon.gypi',`
`npm ERR! gyp info spawn args   '-I',`
`npm ERR! gyp info spawn args   '/root/.cache/node-gyp/16.19.1/include/node/common.gypi',`
`npm ERR! gyp info spawn args   '-Dlibrary=shared_library',`
`npm ERR! gyp info spawn args   '-Dvisibility=default',`
`npm ERR! gyp info spawn args   '-Dnode_root_dir=/root/.cache/node-gyp/16.19.1',`
`npm ERR! gyp info spawn args   '-Dnode_gyp_dir=/usr/local/lib/node_modules/npm/node_modules/node-gyp',`
`npm ERR! gyp info spawn args   '-Dnode_lib_file=/root/.cache/node-gyp/16.19.1/<(target_arch)/node.lib',`
`npm ERR! gyp info spawn args   '-Dmodule_root_dir=/opt/atlassian/pipelines/agent/build/server/node_modules/zeromq',`
`npm ERR! gyp info spawn args   '-Dnode_engine=v8',`
`npm ERR! gyp info spawn args   '--depth=.',`
`npm ERR! gyp info spawn args   '--no-parallel',`
`npm ERR! gyp info spawn args   '--generator-output',`
`npm ERR! gyp info spawn args   'build',`
`npm ERR! gyp info spawn args   '-Goutput_dir=.'`
`npm ERR! gyp info spawn args ]`
`npm ERR! gyp info spawn make`
`npm ERR! gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]`
`npm ERR! /bin/sh: 1: cmake: not found`
`npm ERR! /bin/sh: 1: cmake: not found`
`npm ERR! In file included from ../src/context.h:4,`
`npm ERR!                  from ../src/context.cc:2:`
`npm ERR! ../src/prefix.h:8:10: fatal error: zmq.h: No such file or directory`
`npm ERR!  #include <zmq.h>`
`npm ERR!           ^~~~~~~`
`npm ERR! compilation terminated.`
`npm ERR! make: *** [[zeromq.target.mk:132](http://zeromq.target.mk:132/): Release/obj.target/zeromq/src/context.o] Error 1`
`npm ERR! gyp ERR! build error `
`npm ERR! gyp ERR! stack Error: 'make' failed with exit code: 2`
`npm ERR! gyp ERR! stack     at ChildProcess.onExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:201:23)`
`npm ERR! gyp ERR! stack     at ChildProcess.emit (node:events:513:28)`
`npm ERR! gyp ERR! stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:293:12)`
`npm ERR! gyp ERR! System Linux 5.15.0-1028-aws`
`npm ERR! gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"`
`npm ERR! gyp ERR! cwd /opt/atlassian/pipelines/agent/build/server/node_modules/zeromq`
`npm ERR! gyp ERR! node -v v16.19.1`
`npm ERR! gyp ERR! node-gyp -v v9.1.0`
`npm ERR! gyp ERR! not ok`

I am on the latest node and npm version. I don't run into this locally either so it makes testing hard.

I have same problem installing zeromq on mac.
I try `npm install zeromq@6.0.0-beta.16 --build-from-source`               
but get kinda same errors as @mehiatt does.


`npm ERR! code 1
npm ERR! path /Users/ghiffo/armadio-del-codice/proveJS/casciaro-mammino/middleware-menager-0mq/node_modules/zeromq
npm ERR! command failed
npm ERR! command sh -c -- (shx test -f ./script/build.js || run-s build.js) && cross-env npm_config_build_from_source=true node-gyp-build
npm ERR! dlopen(/Users/ghiffo/armadio-del-codice/proveJS/casciaro-mammino/middleware-menager-0mq/node_modules/zeromq/prebuilds/darwin-arm64/node.napi.glibc.node, 0x0001): Library not loaded: /Users/runner/arm-target/opt/libsodium/lib/libsodium.23.dylib
npm ERR!   Referenced from: <908F1187-CACA-38F3-B226-4E16BB991AD8> /Users/ghiffo/armadio-del-codice/proveJS/casciaro-mammino/middleware-menager-0mq/node_modules/zeromq/prebuilds/darwin-arm64/node.napi.glibc.node
npm ERR!   Reason: tried: '/Users/runner/arm-target/opt/libsodium/lib/libsodium.23.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/runner/arm-target/opt/libsodium/lib/libsodium.23.dylib' (no such file), '/Users/runner/arm-target/opt/libsodium/lib/libsodium.23.dylib' (no such file), '/usr/local/lib/libsodium.23.dylib' (no such file), '/usr/lib/libsodium.23.dylib' (no such file, not in dyld cache)
npm ERR! gyp info it worked if it ends with ok
npm ERR! gyp info using node-gyp@9.1.0
npm ERR! gyp info using node@18.13.0 | darwin | arm64
npm ERR! gyp info find Python using Python version 3.9.4 found at "/Library/Frameworks/Python.framework/Versions/3.9/bin/python3"
npm ERR! gyp info spawn /Library/Frameworks/Python.framework/Versions/3.9/bin/python3
npm ERR! gyp info spawn args [
npm ERR! gyp info spawn args   '/Users/ghiffo/.nvm/versions/node/v18.13.0/lib/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py',
npm ERR! gyp info spawn args   'binding.gyp',
npm ERR! gyp info spawn args   '-f',
npm ERR! gyp info spawn args   'make',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   '/Users/ghiffo/armadio-del-codice/proveJS/casciaro-mammino/middleware-menager-0mq/node_modules/zeromq/build/config.gypi',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   '/Users/ghiffo/.nvm/versions/node/v18.13.0/lib/node_modules/npm/node_modules/node-gyp/addon.gypi',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   '/Users/ghiffo/Library/Caches/node-gyp/18.13.0/include/node/common.gypi',
npm ERR! gyp info spawn args   '-Dlibrary=shared_library',
npm ERR! gyp info spawn args   '-Dvisibility=default',
npm ERR! gyp info spawn args   '-Dnode_root_dir=/Users/ghiffo/Library/Caches/node-gyp/18.13.0',
npm ERR! gyp info spawn args   '-Dnode_gyp_dir=/Users/ghiffo/.nvm/versions/node/v18.13.0/lib/node_modules/npm/node_modules/node-gyp',
npm ERR! gyp info spawn args   '-Dnode_lib_file=/Users/ghiffo/Library/Caches/node-gyp/18.13.0/<(target_arch)/node.lib',
npm ERR! gyp info spawn args   '-Dmodule_root_dir=/Users/ghiffo/armadio-del-codice/proveJS/casciaro-mammino/middleware-menager-0mq/node_modules/zeromq',
npm ERR! gyp info spawn args   '-Dnode_engine=v8',
npm ERR! gyp info spawn args   '--depth=.',
npm ERR! gyp info spawn args   '--no-parallel',
npm ERR! gyp info spawn args   '--generator-output',
npm ERR! gyp info spawn args   'build',
npm ERR! gyp info spawn args   '-Goutput_dir=.'
npm ERR! gyp info spawn args ]
npm ERR! /bin/sh: pkg-config: command not found
npm ERR! gyp: Call to 'pkg-config libsodium --libs' returned exit status 127 while in binding.gyp. while trying to load binding.gyp
npm ERR! gyp ERR! configure error 
npm ERR! gyp ERR! stack Error: `gyp` failed with exit code: 1
npm ERR! gyp ERR! stack     at ChildProcess.onCpExit (/Users/ghiffo/.nvm/versions/node/v18.13.0/lib/node_modules/npm/node_modules/node-gyp/lib/configure.js:284:16)
npm ERR! gyp ERR! stack     at ChildProcess.emit (node:events:513:28)
npm ERR! gyp ERR! stack     at ChildProcess._handle.onexit (node:internal/child_process:291:12)
npm ERR! gyp ERR! System Darwin 22.3.0
npm ERR! gyp ERR! command "/Users/ghiffo/.nvm/versions/node/v18.13.0/bin/node" "/Users/ghiffo/.nvm/versions/node/v18.13.0/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
npm ERR! gyp ERR! cwd /Users/ghiffo/armadio-del-codice/proveJS/casciaro-mammino/middleware-menager-0mq/node_modules/zeromq
npm ERR! gyp ERR! node -v v18.13.0
npm ERR! gyp ERR! node-gyp -v v9.1.0
npm ERR! gyp ERR! not ok

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/ghiffo/.npm/_logs/2023-02-28T23_22_39_163Z-debug-0.log`

--- a/.github/workflows/CI.yml
+++ b/.github/workflows/CI.yml
@@ -170,7 +170,7 @@ jobs:
         run: |
           EOF=$(dd if=/dev/urandom bs=15 count=1 status=none | base64)
           echo "install_deps<<$EOF" >> $GITHUB_OUTPUT
-          cat ./script/install-deps.bash >> $GITHUB_OUTPUT
+          cat ./script/install-deps.sh >> $GITHUB_OUTPUT
           echo "$EOF" >> $GITHUB_OUTPUT
 
       - name: Build Linux Arm64

--- a/.github/workflows/CI.yml
+++ b/.github/workflows/CI.yml
@@ -9,6 +9,7 @@ jobs:
   Build:
     if: ${{ !contains(github.event.head_commit.message, '[skip ci]') }}
     runs-on: ${{ matrix.os }}
+    name: ${{ matrix.os }}-${{ matrix.node_arch }}-${{ matrix.dockerfile }}-${{ matrix.distro }}
     strategy:
       fail-fast: false
       matrix:
@@ -47,16 +48,56 @@ jobs:
             cpp_arch: x64
 
           # Debian Arm
-          - os: macos-14
+          - os: ubuntu-22.04
             node_arch: arm64
             cpp_arch: amd64_arm64
-            dockerfile: docker/alpine.dockerfile
+            distro: node:18
+            install: |
+              apt-get update -y && \
+                  apt-get install --no-install-recommends -y \
+                  bash \
+                  build-essential \
+                  curl \
+                  git \
+                  g++ \
+                  make \
+                  ninja-build \
+                  pkg-config \
+                  unzip \
+                  zip \
+                  python3 \
+                  tar \
+                  cmake \
+                  ninja-build \
+                  automake \
+                  autoconf \
+                  libtool
 
           # Musl Alpine Arm
-          - os: macos-14
+          - os: ubuntu-22.04
             node_arch: arm64
             cpp_arch: amd64_arm64
-            dockerfile: docker/debian.dockerfile
+            install: |
+              apk add --no-cache \
+                  bash \ 
+                  build-base \
+                  curl \ 
+                  git \
+                  g++ \ 
+                  make \ 
+                  ninja-build \
+                  pkgconfig \
+                  unzip \
+                  zip \
+                  python3 \
+                  tar \
+                  cmake \ 
+                  ninja \
+                  musl-dev \
+                  automake \
+                  autoconf \
+                  libtool && \
+                  cp /usr/lib/ninja-build/bin/ninja /usr/bin/ninja
 
     env:
       npm_config_arch: ${{ matrix.node_arch }}
@@ -85,7 +126,7 @@ jobs:
         shell: bash
 
       - name: Setup Cpp
-        if: ${{ !matrix.dockerfile }}
+        if: ${{ !matrix.dockerfile && !matrix.install }}
         uses: aminya/setup-cpp@v1
         with:
           vcvarsall: ${{ contains(matrix.os, 'windows') }}
@@ -101,34 +142,34 @@ jobs:
           brew install gnutls autoconf automake libtool
 
       - uses: pnpm/action-setup@v4
-        if: ${{ !matrix.dockerfile }}
+        if: ${{ !matrix.dockerfile && !matrix.install }}
         with:
           version: 9
 
       - name: Install Node 20
-        if: ${{ !matrix.dockerfile }}
+        if: ${{ !matrix.dockerfile && !matrix.install }}
         uses: actions/setup-node@v4
         with:
           node-version: 20
           architecture: ${{ env.setup_node_arch }}
 
       - name: Install and Build Native
-        if: ${{ !matrix.dockerfile }}
+        if: ${{ !matrix.dockerfile && !matrix.install }}
         run: pnpm install
 
       - name: Build JavaScript
-        if: ${{ !matrix.dockerfile }}
+        if: ${{ !matrix.dockerfile && !matrix.install }}
         run: pnpm run build.js
 
       - name: Install Node 10
-        if: ${{ !matrix.dockerfile && matrix.os != 'macos-14' }}
+        if: ${{ !matrix.dockerfile && !matrix.install && matrix.os != 'macos-14' }}
         uses: actions/setup-node@v4
         with:
           node-version: 10
           architecture: ${{ env.setup_node_arch }}
-
+        
       - name: Build Native
-        if: ${{ !matrix.dockerfile && matrix.node_arch != 'ia32' }}
+        if: ${{ !matrix.dockerfile && !matrix.install && matrix.node_arch != 'ia32' }}
         run: npm run build.native
 
       - name: Build Native Windows 32
@@ -138,7 +179,7 @@ jobs:
           windows-x86
 
       - name: Use Node 20
-        if: ${{ !matrix.dockerfile }}
+        if: ${{ !matrix.dockerfile && !matrix.install }}
         uses: actions/setup-node@v4
         with:
           node-version: 20
@@ -153,6 +194,26 @@ jobs:
           docker cp zeromq-temp:/app/build ./
           docker rm -f zeromq-temp
 
+      - name: Build Linux Arm64
+        if: ${{ matrix.install }} 
+        uses: uraimo/run-on-arch-action@v2
+        with:
+          arch: aarch64
+          distro: ${{ matrix.distro }}
+          githubToken: ${{ github.token }}
+          setup: |
+            mkdir -p "${PWD}/build"
+          dockerRunArgs: |
+            --volume "${PWD}/build:/build"
+          env: |
+            VCPKG_FORCE_SYSTEM_BINARIES: 1
+          install: |
+            ${{ matrix.install }}
+          run: |
+            npm i -g pnpm && \
+            pnpm install && \
+            pnpm run build
+
       - name: Upload build
         uses: actions/upload-artifact@v4
         with:
@@ -161,11 +222,11 @@ jobs:
           overwrite: true
 
       - name: Lint
-        if: "${{ contains(matrix.os, 'ubuntu') && !matrix.dockerfile }}"
+        if: "${{ contains(matrix.os, 'ubuntu') && !matrix.dockerfile && !matrix.install }}"
         run: pnpm run lint-test
 
       - name: Test
-        if: ${{ !matrix.dockerfile }}
+        if: ${{ !matrix.dockerfile && !matrix.install }}
         uses: nick-fields/retry@v3
         with:
           timeout_minutes: 5
@@ -176,7 +237,7 @@ jobs:
             rm -rf ./tmp && mkdir -p ./tmp
 
       - name: Test Electron Windows/MacOS
-        if: "${{ !contains(matrix.os, 'ubuntu') && !matrix.dockerfile }}"
+        if: "${{ !contains(matrix.os, 'ubuntu') && !matrix.dockerfile && !matrix.install }}"
         uses: nick-fields/retry@v3
         with:
           timeout_minutes: 5
@@ -186,7 +247,7 @@ jobs:
         continue-on-error: true
 
       - name: Test Electron Linux
-        if: "${{ contains(matrix.os, 'ubuntu') && !matrix.dockerfile }}"
+        if: "${{ contains(matrix.os, 'ubuntu') && !matrix.dockerfile && !matrix.install }}"
         uses: nick-fields/retry@v3
         with:
           timeout_minutes: 5

--- a/.github/workflows/CI.yml
+++ b/.github/workflows/CI.yml
@@ -60,7 +60,7 @@ jobs:
           - os: ubuntu-24.04
             node_arch: arm64
             cpp_arch: amd64_arm64
-            distro: bullseye
+            distro: bookworm
             native: false
 
           # Musl Alpine Arm
@@ -178,6 +178,7 @@ jobs:
           env: |
             VCPKG_FORCE_SYSTEM_BINARIES: 1
           install: |
+            set -x
             case "${{ matrix.distro }}" in
                 ubuntu*|jessie|stretch|buster|bullseye)
                   apt-get update -q -y
@@ -225,6 +226,7 @@ jobs:
                      nodejs
                  ;;
                 alpine*)
+                  apk update
                   apk add --no-cache \
                     bash \ 
                     build-base \

--- a/.github/workflows/CI.yml
+++ b/.github/workflows/CI.yml
@@ -180,29 +180,6 @@ jobs:
           install: |
             set -x
             case "${{ matrix.distro }}" in
-                ubuntu*|jessie|stretch|buster|bullseye)
-                  apt-get update -q -y
-                  apt-get install --no-install-recommends -y \
-                     bash \
-                     build-essential \
-                     curl \
-                     git \
-                     g++ \
-                     make \
-                     ninja-build \
-                     pkg-config \
-                     unzip \
-                     zip \
-                     python3 \
-                     tar \
-                     cmake \
-                     ninja-build \
-                     automake \
-                     autoconf \
-                     libtool \
-                     nodejs \
-                     npm
-                 ;;
                 fedora*)
                   dnf update -q -y
                   dnf install -y \
@@ -227,27 +204,32 @@ jobs:
                  ;;
                 alpine*)
                   apk update
-                  apk add --no-cache \
-                    bash \ 
-                    build-base \
-                    curl \ 
-                    git \
-                    g++ \ 
-                    make \ 
-                    ninja-build \
-                    pkgconfig \
-                    unzip \
-                    zip \
-                    python3 \
-                    tar \
-                    cmake \ 
-                    musl-dev \
-                    automake \
-                    autoconf \
-                    libtool \
-                    nodejs \
-                    npm && \
-                    cp /usr/lib/ninja-build/bin/ninja /usr/bin/ninja
+                  apk add --no-cache bash build-base curl git g++ make ninja-build pkgconfig unzip zip python3 tar cmake musl-dev automake autoconf libtool nodejs npm 
+                  cp /usr/lib/ninja-build/bin/ninja /usr/bin/ninja
+                 ;;
+                *)
+                  # Default case (Ubuntu/Debian)
+                  apt-get update -q -y
+                  apt-get install --no-install-recommends -y \
+                     bash \
+                     build-essential \
+                     curl \
+                     git \
+                     g++ \
+                     make \
+                     ninja-build \
+                     pkg-config \
+                     unzip \
+                     zip \
+                     python3 \
+                     tar \
+                     cmake \
+                     ninja-build \
+                     automake \
+                     autoconf \
+                     libtool \
+                     nodejs \
+                     npm
                  ;;
              esac
              npm i -g pnpm

--- a/.github/workflows/CI.yml
+++ b/.github/workflows/CI.yml
@@ -9,6 +9,7 @@ jobs:
   Build:
     if: ${{ !contains(github.event.head_commit.message, '[skip ci]') }}
     runs-on: ${{ matrix.os }}
+    # prettier-ignore
     name: ${{ matrix.os }}-${{ matrix.node_arch }}-${{ matrix.dockerfile }}-${{ matrix.distro }}
     strategy:
       fail-fast: false
@@ -22,11 +23,16 @@ jobs:
           - x64
         dockerfile:
           - ""
+        distro:
+          - ""
+        native:
+          - true
 
         include:
           - os: windows-2019
             node_arch: ia32
             cpp_arch: amd64_x86
+            native: true
 
           # - os: windows-2022
           #   node_arch: x64
@@ -36,68 +42,33 @@ jobs:
           - os: macos-13
             node_arch: x64
             cpp_arch: x64
+            native: true
 
           - os: macos-14
             node_arch: arm64
             cpp_arch: amd64_arm64
+            native: true
 
           # Musl Alpine
           - os: ubuntu-22.04
             dockerfile: docker/alpine.dockerfile
             node_arch: x64
             cpp_arch: x64
+            native: false
 
           # Debian Arm
           - os: ubuntu-22.04
             node_arch: arm64
             cpp_arch: amd64_arm64
-            distro: node:18
-            install: |
-              apt-get update -y && \
-                  apt-get install --no-install-recommends -y \
-                  bash \
-                  build-essential \
-                  curl \
-                  git \
-                  g++ \
-                  make \
-                  ninja-build \
-                  pkg-config \
-                  unzip \
-                  zip \
-                  python3 \
-                  tar \
-                  cmake \
-                  ninja-build \
-                  automake \
-                  autoconf \
-                  libtool
+            distro: bullseye
+            native: false
 
           # Musl Alpine Arm
           - os: ubuntu-22.04
             node_arch: arm64
             cpp_arch: amd64_arm64
-            install: |
-              apk add --no-cache \
-                  bash \ 
-                  build-base \
-                  curl \ 
-                  git \
-                  g++ \ 
-                  make \ 
-                  ninja-build \
-                  pkgconfig \
-                  unzip \
-                  zip \
-                  python3 \
-                  tar \
-                  cmake \ 
-                  ninja \
-                  musl-dev \
-                  automake \
-                  autoconf \
-                  libtool && \
-                  cp /usr/lib/ninja-build/bin/ninja /usr/bin/ninja
+            distro: alpine_latest
+            native: false
 
     env:
       npm_config_arch: ${{ matrix.node_arch }}
@@ -112,9 +83,8 @@ jobs:
           path: |
             ./node_modules/
             ~/vcpkg
-          key:
-            # prettier-ignore
-            "cache-OS:${{ matrix.os }}-arch:${{ matrix.node_arch}}-${{hashFiles('./pnpm-lock.yaml', './vcpkg.json', './CMakeLists.txt') }}"
+          # prettier-ignore
+          key: "cache-OS:${{ matrix.os }}-arch:${{ matrix.node_arch}}-${{hashFiles('./pnpm-lock.yaml', './vcpkg.json', './CMakeLists.txt') }}"
           restore-keys: |
             "cache-OS:${{ matrix.os }}-arch:${{ matrix.node_arch }}-"
 
@@ -126,7 +96,7 @@ jobs:
         shell: bash
 
       - name: Setup Cpp
-        if: ${{ !matrix.dockerfile && !matrix.install }}
+        if: ${{ matrix.native }}
         uses: aminya/setup-cpp@v1
         with:
           vcvarsall: ${{ contains(matrix.os, 'windows') }}
@@ -142,34 +112,34 @@ jobs:
           brew install gnutls autoconf automake libtool
 
       - uses: pnpm/action-setup@v4
-        if: ${{ !matrix.dockerfile && !matrix.install }}
+        if: ${{ matrix.native }}
         with:
           version: 9
 
       - name: Install Node 20
-        if: ${{ !matrix.dockerfile && !matrix.install }}
+        if: ${{ matrix.native }}
         uses: actions/setup-node@v4
         with:
           node-version: 20
           architecture: ${{ env.setup_node_arch }}
 
       - name: Install and Build Native
-        if: ${{ !matrix.dockerfile && !matrix.install }}
+        if: ${{ matrix.native }}
         run: pnpm install
 
       - name: Build JavaScript
-        if: ${{ !matrix.dockerfile && !matrix.install }}
+        if: ${{ matrix.native }}
         run: pnpm run build.js
 
       - name: Install Node 10
-        if: ${{ !matrix.dockerfile && !matrix.install && matrix.os != 'macos-14' }}
+        if: ${{ matrix.native && matrix.os != 'macos-14' }}
         uses: actions/setup-node@v4
         with:
           node-version: 10
           architecture: ${{ env.setup_node_arch }}
-        
+
       - name: Build Native
-        if: ${{ !matrix.dockerfile && !matrix.install && matrix.node_arch != 'ia32' }}
+        if: ${{ matrix.native && matrix.node_arch != 'ia32' }}
         run: npm run build.native
 
       - name: Build Native Windows 32
@@ -179,7 +149,7 @@ jobs:
           windows-x86
 
       - name: Use Node 20
-        if: ${{ !matrix.dockerfile && !matrix.install }}
+        if: ${{ matrix.native }}
         uses: actions/setup-node@v4
         with:
           node-version: 20
@@ -195,7 +165,7 @@ jobs:
           docker rm -f zeromq-temp
 
       - name: Build Linux Arm64
-        if: ${{ matrix.install }} 
+        if: ${{ matrix.distro }}
         uses: uraimo/run-on-arch-action@v2
         with:
           arch: aarch64
@@ -208,7 +178,76 @@ jobs:
           env: |
             VCPKG_FORCE_SYSTEM_BINARIES: 1
           install: |
-            ${{ matrix.install }}
+            case "${{ matrix.distro }}" in
+                ubuntu*|jessie|stretch|buster|bullseye)
+                  apt-get update -q -y
+                  apt-get install --no-install-recommends -y \
+                     bash \
+                     build-essential \
+                     curl \
+                     git \
+                     g++ \
+                     make \
+                     ninja-build \
+                     pkg-config \
+                     unzip \
+                     zip \
+                     python3 \
+                     tar \
+                     cmake \
+                     ninja-build \
+                     automake \
+                     autoconf \
+                     libtool \
+                     nodejs
+                 ;;
+                fedora*)
+                  dnf update -q -y
+                  dnf install -y \
+                     bash \
+                     build-essential \
+                     curl \
+                     git \
+                     g++ \
+                     make \
+                     ninja-build \
+                     pkg-config \
+                     unzip \
+                     zip \
+                     python3 \
+                     tar \
+                     cmake \
+                     ninja-build \
+                     automake \
+                     autoconf \
+                     libtool \
+                     nodejs
+                 ;;
+                alpine*)
+                  apk add --no-cache \
+                    bash \ 
+                    build-base \
+                    curl \ 
+                    git \
+                    g++ \ 
+                    make \ 
+                    ninja-build \
+                    pkgconfig \
+                    unzip \
+                    zip \
+                    python3 \
+                    tar \
+                    cmake \ 
+                    ninja \
+                    musl-dev \
+                    automake \
+                    autoconf \
+                    libtool \
+                    nodejs && \
+                    cp /usr/lib/ninja-build/bin/ninja /usr/bin/ninja
+                 ;;
+             esac
+
           run: |
             npm i -g pnpm && \
             pnpm install && \
@@ -222,11 +261,11 @@ jobs:
           overwrite: true
 
       - name: Lint
-        if: "${{ contains(matrix.os, 'ubuntu') && !matrix.dockerfile && !matrix.install }}"
+        if: "${{ contains(matrix.os, 'ubuntu') && matrix.native }}"
         run: pnpm run lint-test
 
       - name: Test
-        if: ${{ !matrix.dockerfile && !matrix.install }}
+        if: ${{ matrix.native }}
         uses: nick-fields/retry@v3
         with:
           timeout_minutes: 5
@@ -237,7 +276,7 @@ jobs:
             rm -rf ./tmp && mkdir -p ./tmp
 
       - name: Test Electron Windows/MacOS
-        if: "${{ !contains(matrix.os, 'ubuntu') && !matrix.dockerfile && !matrix.install }}"
+        if: "${{ !contains(matrix.os, 'ubuntu') && matrix.native }}"
         uses: nick-fields/retry@v3
         with:
           timeout_minutes: 5
@@ -247,7 +286,7 @@ jobs:
         continue-on-error: true
 
       - name: Test Electron Linux
-        if: "${{ contains(matrix.os, 'ubuntu') && !matrix.dockerfile && !matrix.install }}"
+        if: "${{ contains(matrix.os, 'ubuntu') && matrix.native }}"
         uses: nick-fields/retry@v3
         with:
           timeout_minutes: 5

--- a/.github/workflows/CI.yml
+++ b/.github/workflows/CI.yml
@@ -178,63 +178,7 @@ jobs:
           env: |
             VCPKG_FORCE_SYSTEM_BINARIES: 1
           install: |
-            set -x
-            case "${{ matrix.distro }}" in
-                fedora*)
-                  dnf update -q -y
-                  dnf install -y \
-                     bash \
-                     build-essential \
-                     curl \
-                     git \
-                     g++ \
-                     make \
-                     ninja-build \
-                     pkg-config \
-                     unzip \
-                     zip \
-                     python3 \
-                     tar \
-                     cmake \
-                     ninja-build \
-                     automake \
-                     autoconf \
-                     libtool \
-                     nodejs
-                 ;;
-                alpine*)
-                  apk update
-                  apk add --no-cache bash build-base curl git g++ make ninja-build pkgconfig unzip zip python3 tar cmake musl-dev automake autoconf libtool nodejs npm 
-                  cp /usr/lib/ninja-build/bin/ninja /usr/bin/ninja
-                 ;;
-                *)
-                  # Default case (Ubuntu/Debian)
-                  apt-get update -q -y
-                  apt-get install --no-install-recommends -y \
-                     bash \
-                     build-essential \
-                     curl \
-                     git \
-                     g++ \
-                     make \
-                     ninja-build \
-                     pkg-config \
-                     unzip \
-                     zip \
-                     python3 \
-                     tar \
-                     cmake \
-                     ninja-build \
-                     automake \
-                     autoconf \
-                     libtool \
-                     nodejs \
-                     npm
-                 ;;
-             esac
-             npm i -g pnpm
-             npx -y setup-cpp --vcpkg "ee2d2a100103e0f3613c60655dcf15be7d5157b8"
-             ~/vcpkg/vcpkg install 'zeromq[draft,curve,sodium]'
+            ./script/install-deps.bash
 
           run: |
             pnpm install && \

--- a/.github/workflows/CI.yml
+++ b/.github/workflows/CI.yml
@@ -233,6 +233,8 @@ jobs:
                  ;;
              esac
              npm i -g pnpm
+             npx -y setup-cpp --vcpkg "ee2d2a100103e0f3613c60655dcf15be7d5157b8"
+             ~/vcpkg/vcpkg install 'zeromq[draft,curve,sodium]'
 
           run: |
             pnpm install && \

--- a/.github/workflows/CI.yml
+++ b/.github/workflows/CI.yml
@@ -168,7 +168,10 @@ jobs:
         if: ${{ matrix.distro }}
         id: read-installer-script
         run: |
-          echo "install_deps=$(cat ./script/install-deps.bash)" >> $GITHUB_OUTPUT
+          EOF=$(dd if=/dev/urandom bs=15 count=1 status=none | base64)
+          echo "install_deps<<$EOF" >> $GITHUB_OUTPUT
+          cat ./script/install-deps.bash >> $GITHUB_OUTPUT
+          echo "$EOF" >> $GITHUB_OUTPUT
 
       - name: Build Linux Arm64
         if: ${{ matrix.distro }}

--- a/.github/workflows/CI.yml
+++ b/.github/workflows/CI.yml
@@ -50,21 +50,21 @@ jobs:
             native: true
 
           # Musl Alpine
-          - os: ubuntu-22.04
+          - os: ubuntu-24.04
             dockerfile: docker/alpine.dockerfile
             node_arch: x64
             cpp_arch: x64
             native: false
 
           # Debian Arm
-          - os: ubuntu-22.04
+          - os: ubuntu-24.04
             node_arch: arm64
             cpp_arch: amd64_arm64
             distro: bullseye
             native: false
 
           # Musl Alpine Arm
-          - os: ubuntu-22.04
+          - os: ubuntu-24.04
             node_arch: arm64
             cpp_arch: amd64_arm64
             distro: alpine_latest
@@ -166,7 +166,7 @@ jobs:
 
       - name: Build Linux Arm64
         if: ${{ matrix.distro }}
-        uses: uraimo/run-on-arch-action@v2
+        uses: uraimo/run-on-arch-action@v2.8.1
         with:
           arch: aarch64
           distro: ${{ matrix.distro }}

--- a/.github/workflows/CI.yml
+++ b/.github/workflows/CI.yml
@@ -199,7 +199,8 @@ jobs:
                      automake \
                      autoconf \
                      libtool \
-                     nodejs
+                     nodejs \
+                     npm
                  ;;
                 fedora*)
                   dnf update -q -y
@@ -238,18 +239,18 @@ jobs:
                     python3 \
                     tar \
                     cmake \ 
-                    ninja \
                     musl-dev \
                     automake \
                     autoconf \
                     libtool \
-                    nodejs && \
+                    nodejs \
+                    npm && \
                     cp /usr/lib/ninja-build/bin/ninja /usr/bin/ninja
                  ;;
              esac
+             npm i -g pnpm
 
           run: |
-            npm i -g pnpm && \
             pnpm install && \
             pnpm run build
 

--- a/.github/workflows/CI.yml
+++ b/.github/workflows/CI.yml
@@ -40,11 +40,23 @@ jobs:
             node_arch: arm64
             cpp_arch: amd64_arm64
 
-          # Alpine
+          # Musl Alpine
           - os: ubuntu-22.04
             dockerfile: docker/alpine.dockerfile
             node_arch: x64
             cpp_arch: x64
+
+          # Debian Arm
+          - os: macos-14
+            node_arch: arm64
+            cpp_arch: amd64_arm64
+            dockerfile: docker/alpine.dockerfile
+
+          # Musl Alpine Arm
+          - os: macos-14
+            node_arch: arm64
+            cpp_arch: amd64_arm64
+            dockerfile: docker/debian.dockerfile
 
     env:
       npm_config_arch: ${{ matrix.node_arch }}
@@ -137,7 +149,8 @@ jobs:
         run: |
           docker build -t zeromq -f ${{ matrix.dockerfile }} .
           docker create --name zeromq-temp zeromq
-          docker cp zeromq-temp:/app/build ./build
+          mkdir -p ./build
+          docker cp zeromq-temp:/app/build ./
           docker rm -f zeromq-temp
 
       - name: Upload build

--- a/.github/workflows/CI.yml
+++ b/.github/workflows/CI.yml
@@ -164,6 +164,12 @@ jobs:
           docker cp zeromq-temp:/app/build ./
           docker rm -f zeromq-temp
 
+      - name: Read Installer Script
+        if: ${{ matrix.distro }}
+        id: read-installer-script
+        run: |
+          echo "install_deps=$(cat ./script/install-deps.bash)" >> $GITHUB_OUTPUT
+
       - name: Build Linux Arm64
         if: ${{ matrix.distro }}
         uses: uraimo/run-on-arch-action@v2.8.1
@@ -178,7 +184,7 @@ jobs:
           env: |
             VCPKG_FORCE_SYSTEM_BINARIES: 1
           install: |
-            ./script/install-deps.bash
+            ${{ steps.read-installer-script.outputs.install_deps }}
 
           run: |
             pnpm install && \

--- a/docker/debian.dockerfile
+++ b/docker/debian.dockerfile
@@ -1,40 +1,37 @@
-FROM node:18-alpine AS builder
+FROM node:18 AS builder
 
 WORKDIR /app
 COPY ./ ./
 ENV VCPKG_FORCE_SYSTEM_BINARIES=1
 RUN \
     # system dependencies
-    apk add --no-cache \
-    bash \ 
-    build-base \
-    curl \ 
+    apt-get update -y && \
+    apt-get install --no-install-recommends -y \
+    bash \
+    build-essential \
+    curl \
     git \
-    g++ \ 
-    make \ 
+    g++ \
+    make \
     ninja-build \
-    pkgconfig \
+    pkg-config \
     unzip \
     zip \
     python3 \
     tar \
-    cmake \ 
-    ninja \
-    musl-dev \
+    cmake \
+    ninja-build \
     automake \
     autoconf \
     libtool && \
-    cp /usr/lib/ninja-build/bin/ninja /usr/bin/ninja && \ 
     # build
     npm i -g pnpm && \
     pnpm install && \
     pnpm run build
 
-FROM node:18-alpine
+FROM node:18
 WORKDIR /app
 
 COPY ./ ./
 COPY --from=builder /app/build ./build
 COPY --from=builder /app/node_modules ./node_modules
-
-

--- a/script/install-deps.bash
+++ b/script/install-deps.bash
@@ -1,0 +1,74 @@
+#!/bin/bash
+
+set -x
+
+# Ubuntu/Debian
+apt=$(command -v apt-get || true)
+if [[ -n $apt ]]; then
+    apt-get update -q -y
+    apt-get install --no-install-recommends -y \
+        bash \
+        build-essential \
+        curl \
+        git \
+        g++ \
+        make \
+        ninja-build \
+        pkg-config \
+        unzip \
+        zip \
+        python3 \
+        tar \
+        cmake \
+        ninja-build \
+        automake \
+        autoconf \
+        libtool \
+        nodejs \
+        npm
+fi
+
+# Alpine Linux
+apk=$(command -v apk || true)
+if [[ -n $apk ]]; then
+    apk update
+    apk add --no-cache bash build-base curl git g++ make ninja-build pkgconfig unzip zip python3 tar cmake musl-dev automake autoconf libtool nodejs npm
+    cp /usr/lib/ninja-build/bin/ninja /usr/bin/ninja
+fi
+
+# Fedora/RHEL
+dnf=$(command -v dnf || true)
+if [[ -n $dnf ]]; then
+    dnf update -q -y
+    dnf install -y \
+        bash \
+        build-essential \
+        curl \
+        git \
+        g++ \
+        make \
+        ninja-build \
+        pkg-config \
+        unzip \
+        zip \
+        python3 \
+        tar \
+        cmake \
+        ninja-build \
+        automake \
+        autoconf \
+        libtool \
+        nodejs
+fi
+
+# pnpm
+npm i -g pnpm
+
+# vcpkg
+npx -y setup-cpp --vcpkg "ee2d2a100103e0f3613c60655dcf15be7d5157b8"
+
+# zeromq
+export VCPKG_FORCE_SYSTEM_BINARIES=1
+cd ~/vcpkg || exit 1
+~/vcpkg/vcpkg install 'zeromq[draft,curve,sodium]' || (cd - || exit 1)
+cd - || exit 1

--- a/script/install-deps.sh
+++ b/script/install-deps.sh
@@ -67,7 +67,7 @@ npm i -g pnpm
 export VCPKG_FORCE_SYSTEM_BINARIES=1
 
 # vcpkg
-npx -y setup-cpp --vcpkg "ee2d2a100103e0f3613c60655dcf15be7d5157b8"
+npx -y setup-cpp --vcpkg "ee2d2a100103e0f3613c60655dcf15be7d5157b8" --timeout 3600000
 
 # zeromq
 cd ~/vcpkg || exit 1

--- a/script/install-deps.sh
+++ b/script/install-deps.sh
@@ -67,7 +67,11 @@ npm i -g pnpm
 export VCPKG_FORCE_SYSTEM_BINARIES=1
 
 # vcpkg
-npx -y setup-cpp --vcpkg "ee2d2a100103e0f3613c60655dcf15be7d5157b8" --timeout 3600000
+git clone https://github.com/microsoft/vcpkg.git ~/vcpkg
+cd ~/vcpkg || exit 1
+git checkout "ee2d2a100103e0f3613c60655dcf15be7d5157b8"
+~/vcpkg/bootstrap-vcpkg.sh
+cd - || exit 1
 
 # zeromq
 cd ~/vcpkg || exit 1

--- a/script/install-deps.sh
+++ b/script/install-deps.sh
@@ -64,11 +64,12 @@ fi
 # pnpm
 npm i -g pnpm
 
+export VCPKG_FORCE_SYSTEM_BINARIES=1
+
 # vcpkg
 npx -y setup-cpp --vcpkg "ee2d2a100103e0f3613c60655dcf15be7d5157b8"
 
 # zeromq
-export VCPKG_FORCE_SYSTEM_BINARIES=1
 cd ~/vcpkg || exit 1
 ~/vcpkg/vcpkg install 'zeromq[draft,curve,sodium]' || (cd - || exit 1)
 cd - || exit 1

--- a/script/install-deps.sh
+++ b/script/install-deps.sh
@@ -1,10 +1,10 @@
-#!/bin/bash
+#!/bin/sh
 
 set -x
 
 # Ubuntu/Debian
 apt=$(command -v apt-get || true)
-if [[ -n $apt ]]; then
+if [ -n "$apt" ]; then
     apt-get update -q -y
     apt-get install --no-install-recommends -y \
         bash \
@@ -30,7 +30,7 @@ fi
 
 # Alpine Linux
 apk=$(command -v apk || true)
-if [[ -n $apk ]]; then
+if [ -n "$apk" ]; then
     apk update
     apk add --no-cache bash build-base curl git g++ make ninja-build pkgconfig unzip zip python3 tar cmake musl-dev automake autoconf libtool nodejs npm
     cp /usr/lib/ninja-build/bin/ninja /usr/bin/ninja
@@ -38,7 +38,7 @@ fi
 
 # Fedora/RHEL
 dnf=$(command -v dnf || true)
-if [[ -n $dnf ]]; then
+if [ -n "$dnf" ]; then
     dnf update -q -y
     dnf install -y \
         bash \

